internal/catalogmetadata/cache/cache.go

```diff
@@ -117,19 +65,35 @@ func (fsc *filesystemCache) FetchCatalogContents(ctx context.Context, catalog *c
 	// updating this, has no way to tell if the current ref is the
 	// newest possible ref. If another thread has already updated
 	// this to be the same value, skip the write logic and return
-	// the cached contents
-	if data, ok := fsc.cacheDataByCatalogName[catalog.Name]; ok {
-		if data.ResolvedRef == catalog.Status.ResolvedSource.Image.ResolvedRef {
-			return os.DirFS(cacheDir), nil
-		}
+	// the cached contents.
+	if cache, err := fsc.get(catalogName, resolvedRef); err == nil && cache != nil {
+		// We only return here if the was no error during
+		// the previous (likely concurrent) cache population attempt.
+		// If there was an error - we want to try and populate the cache again.
+		return cache, nil
+	}
```

Comment:
Apologies for missing this the first time around, but do we actually still need this?

I think this was put in place with more complex thread safety logic than we require now since we are doing thread-safe "Put" and "Get" operations. Refreshing my memory on the previous logic, it was done for the specific scenario of:
- We do a read-only lock to verify if the content has been cached before
- We release the read-only lock once we have verified that either we have or have not cached this content before
- Some other thread could now potentially get a write lock and write the contents for this catalog to the cache
- When we go to write content for the catalog to the cache, we get a write lock and re-verify whether or not it was written by some other process. We let it win because we don't know whether our revision of the content is newer or not.

I don't think the same scenario is possible _here_ since we are doing a lock for the entire duration of the `Put` method and as such we can probably discard this initial get and ensure `Put` has the single responsibility of _always_ writing the new cache.

Reply:
That being said, it may still be valid to say "we have successfully cached this revision before so we will ignore a failed put for the same revision". I would lean more towards putting the responsibility of this verification to the caller though to simplify the responsibility of these cache functions.

Reply:
@everettraven it is still possible that two callers still do the following concurrently:

1. Get cache - it is missing - we lock for reading
2. Do an HTTP request - we do not lock this
3. Write into FS - we lock this

So if this happens we have and we remove the get from `Put` we have two redundant things: 1) network request 2) FS IO.

So if we remove this check we are risking two sequential writes to FS which will make reads blocked as well. With this check we ensure that only one FS write happens even if there was multiple concurrent attempts to fetch cache.

Reply:
Right. I get that it would still be possible for that scenario to occur, but I'm questioning whether it should be the responsibility of the `Put` method to handle that or the caller. I would think the caller should take that into consideration as opposed to the `Put` method.

Reply:
@everettraven if we make this responsibility of the caller then the caller will have to be responsible for locking of the mutex as well. And we probably don't want to make locking the caller's responsibility.

Or am I still not getting the idea?